### PR TITLE
#514: parquet marshalling

### PR DESCRIPTION
**Pull request by efgefg0001**

### CODE REVIEW

File: server/schema/processed_file.go [hunk 1]
```diff
@@ -32,6 +32,14 @@ func (pf *ProcessedFile) GetPayloadBytes(marshaller Marshaller) []byte {
 	return b
 }
 
+//GetPayloadUsingStronglyTypedMarshaller returns bytes, containing marshalled payload
+//StronglyTypedMarshaller needs to know payload schema (types of fields) to convert payload to byte slice
+func (pf *ProcessedFile) GetPayloadUsingStronglyTypedMarshaller(stm StronglyTypedMarshaller) []byte {
+	b, _ := stm.Marshal(pf.BatchHeader, pf.payload)
+	//TODO do i need error handling here?
```
Comment (xtreding) [inline server/schema/processed_file.go:39]: let's return an error if occurred. Handle this error in storages

File: server/schema/strongly_typed_marshaller.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+			md = append(md, fmt.Sprintf("name=%s, type=BOOLEAN", field))
+		case typing.INT64:
+			md = append(md, fmt.Sprintf("name=%s, type=INT64", field))
+		case typing.FLOAT64:
+			md = append(md, fmt.Sprintf("name=%s, type=DOUBLE", field))
+		case typing.STRING:
+			md = append(md, fmt.Sprintf("name=%s, type=BYTE_ARRAY, convertedtype=UTF8, encoding=PLAIN_DICTIONARY", field))
+		case typing.TIMESTAMP:
+			md = append(md, fmt.Sprintf("name=%s, type=TIMESTAMP_MILLIS", field))
+		// UNKNOWN and default
```
Comment (xtreding) [inline server/schema/strongly_typed_marshaller.go:69]: let's return an error if unknown type

File: server/schema/strongly_typed_marshaller.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+		case typing.BOOL, typing.INT64, typing.FLOAT64, typing.STRING:
+			rec[index] = fieldValue
+		case typing.TIMESTAMP:
+			switch fieldValue := fieldValue.(type) {
+			case time.Time:
+				rec[index] = types.TimeToTIMESTAMP_MILLIS(fieldValue, true)
+			default:
+				rec[index] = types.TimeToTIMESTAMP_MILLIS(time.Now(), true)
+			}
+		// UNKNOWN and default
```
Comment (xtreding) [inline server/schema/strongly_typed_marshaller.go:96]: the same

File: server/storages/s3.go [hunk 4]
```diff
@@ -116,11 +115,17 @@ func (s3 *S3) Store(fileName string, objects []map[string]interface{}, alreadyUp
...
-		return schema.JSONMarshallerInstance
+func (s3 *S3) marshall(fdata *schema.ProcessedFile) []byte {
+	switch s3.s3Adapter.Format() {
+	case adapters.S3FormatCSV:
+		return fdata.GetPayloadBytes(schema.CSVMarshallerInstance)
+	case adapters.S3FormatFlatJSON, adapters.S3FormatJSON:
+		return fdata.GetPayloadBytes(schema.JSONMarshallerInstance)
+	case adapters.S3FormatParquet:
+		return fdata.GetPayloadUsingStronglyTypedMarshaller(schema.NewParquetMarshaller())
+	default:
```
Comment (xtreding) [inline server/storages/s3.go:126]: default error

File: server/storages/s3.go [hunk 5]
```diff
@@ -80,9 +80,8 @@ func (s3 *S3) Store(fileName string, objects []map[string]interface{}, alreadyUp
 
 	storeFailedEvents := true
 	tableResults := map[string]*StoreResult{}
-	marshaller := s3.marshaller()
 	for _, fdata := range processedFiles {
-		b := fdata.GetPayloadBytes(marshaller)
+		b := s3.marshall(fdata)
```
Comment (xtreding) [inline server/storages/s3.go:84]: let's refactor this code to:
1. create a func like: func load (f *schema.ProcessedFile) error{
                b,err := s3.marshall(fdata)
                if err != nil { 
                  return err
                }
		fileName := s3.fileName(fdata)
		return s3.s3Adapter.UploadBytes(fileName, b)
}

2. replace this code with:
for _, fdata := range processedFiles {
err := s3.load(fdata)
tableResults[fdata.BatchHeader.TableName] = &StoreResult{Err: err, RowsCount: fdata.GetPayloadLen(), EventsSrc: fdata.GetEventsPerSrc()}
		if err != nil {
			logging.Errorf("[%s] Error storing file %s: %v", s3.ID(), fileName, err)
			storeFailedEvents = false
		}
		//other code as is
		....

}

**Review by xtreding — CHANGES_REQUESTED** [review]

File: server/schema/processed_file.go (hunk 1 above)
Comment (efgefg0001) [reply server/schema/processed_file.go:39]: done

File: server/schema/strongly_typed_marshaller.go (hunk 2 above)
Comment (efgefg0001) [reply server/schema/strongly_typed_marshaller.go]: done

File: server/schema/strongly_typed_marshaller.go (hunk 3 above)
Comment (efgefg0001) [reply server/schema/strongly_typed_marshaller.go]: done

File: server/storages/s3.go (hunk 4 above)
Comment (efgefg0001) [reply server/storages/s3.go:126]: done

File: server/storages/s3.go (hunk 5 above)
Comment (efgefg0001) [reply server/storages/s3.go]: done

**Review by xtreding — APPROVED** [review]

**Pull request merged**